crates/autopilot/src/infra/api.rs

```diff
@@ -0,0 +1,94 @@
+use {
+    alloy::primitives::Address,
+    axum::{
+        Router,
+        extract::{Path, Query, State as AxumState},
+        http::StatusCode,
+        response::{IntoResponse, Json, Response},
+        routing::get,
+    },
+    model::quote::NativeTokenPrice,
+    observe::distributed_tracing::tracing_axum::{make_span, record_trace_id},
+    serde::Deserialize,
+    shared::price_estimation::{PriceEstimationError, native::NativePriceEstimating},
+    std::{net::SocketAddr, sync::Arc, time::Duration},
+    tokio::sync::oneshot,
+};
+
+#[derive(Clone)]
+struct State {
+    estimator: Arc<dyn NativePriceEstimating>,
+    timeout: Duration,
+}
+
+#[derive(Debug, Deserialize)]
+struct NativePriceQuery {
+    /// Optional timeout in milliseconds for the price estimation request.
+    /// If not provided, uses the default timeout configured for autopilot.
+    #[serde(default)]
+    timeout_ms: Option<u64>,
+}
+
+pub async fn serve(
+    addr: SocketAddr,
+    estimator: Arc<dyn NativePriceEstimating>,
+    timeout: Duration,
+    shutdown: oneshot::Receiver<()>,
+) -> Result<(), hyper::Error> {
+    let state = State { estimator, timeout };
+
+    let app = Router::new()
+        .route("/native_price/:token", get(get_native_price))
+        .with_state(state)
+        .layer(
+            tower::ServiceBuilder::new()
+                .layer(tower_http::trace::TraceLayer::new_for_http().make_span_with(make_span))
+                .map_request(record_trace_id),
+        );
+
+    let server = axum::Server::bind(&addr).serve(app.into_make_service());
+    tracing::info!(?addr, "serving native price API");
+
+    server
+        .with_graceful_shutdown(async {
+            shutdown.await.ok();
+        })
+        .await
+}
+
+async fn get_native_price(
+    Path(token): Path<Address>,
+    Query(query): Query<NativePriceQuery>,
+    AxumState(state): AxumState<State>,
+) -> Response {
+    let timeout = query
+        .timeout_ms
+        .map(Duration::from_millis)
+        .unwrap_or(state.timeout);
```

Comment:
We should be defensive and ensure that the timeout is at most our configured default timeout.